plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -734,49 +882,86 @@ public void finishInsertTable(ConnectorSession session, JdbcOutputTableHandle ha
                 Optional.ofNullable(handle.getCatalogName()),
                 Optional.ofNullable(handle.getSchemaName()),
                 handle.getTableName());
-        String insertSql = buildInsertSql(session, targetTable, temporaryTable, handle.getColumnNames());
+
+        // We conditionally create more than the one table, so keep a list of the tables that need to be dropped.
+        List<RemoteTableName> droppedAfter = new ArrayList<>();
+        droppedAfter.add(temporaryTable);
 
         try (Connection connection = getConnection(session, handle)) {
+            String columns = handle.getColumnNames().stream()
+                    .map(this::quoted)
+                    .collect(joining(", "));
+
+            String insertSql = format("INSERT INTO %s %s (%s) SELECT %s FROM %s x",
+                    quoted(targetTable),
+                    isTableLockNeeded(session) ? "WITH (TABLOCK)" : "", // TABLOCK is a prerequisite for minimal logging in SQL Server
+                    columns,
+                    columns,
+                    quoted(temporaryTable));
+
+            if (includePageSinkIdColumn(session)) {
+                RemoteTableName pageSinkTable = constructHelperTable(session, connection, handle, pageSinkIds);
+                droppedAfter.add(pageSinkTable);
+
+                String pageSinkColumn = identifierMapping.toRemoteColumnName(connection, getPageSinkIdColumnName(handle.getColumnNames()));
+
+                insertSql += format(" WHERE EXISTS (SELECT 1 FROM %s y where y.%s = x.%s)",
+                        quoted(pageSinkTable),
+                        pageSinkColumn,
+                        pageSinkColumn);
+            }
+
             execute(connection, insertSql);
         }
         catch (SQLException e) {
             throw new TrinoException(JDBC_ERROR, e);
         }
         finally {
-            dropTable(session, temporaryTable);
+            droppedAfter.stream()
+                    .forEach(t -> {
+                        try {
+                            dropTable(session, t);
+                        }
+                        catch (Exception e) {
+                            log.error(e, "Exception thrown while dropping table %s", t);
```

Comment:
this changes semantic. Now we will not throw if everthin succeeds but we fail on cleanup.
Maybe use `Closer` instead:

```
        Closer closer = Closer.create();
        closer.register(() -> dropTable(session, temporaryTable));
...
          closer.register(() -> dropTable(session, pageSinkTable));
...
            try {
                closer.close();
            }
            catch (IOException e) {
                throw new RuntimeException(e);
            }
```